### ISSUE TITLE
BC breaking change to routes in 3.8.8

### ISSUE THREAD

**Issue by burzum** [cakephp/cakephp#14325]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.8.8 and newer

### What you did
Updated to 3.8.8

### What happened

Some tests started to fail that check for correct URLs.

```text
1) App\Test\TestCase\View\Helper\SearchFilterHelperTest::testRemoveQueryParam
Cake\Routing\Exception\MissingRouteException: A route matching "array (
  '?' =>
  array (
    'section' => 'profiles',
    'preserve' => 1,
  ),
  'language' => 'en',
  'action' => 'index',
  'plugin' => NULL,
  'controller' => NULL,
  '_ext' => NULL,
)" could not be found.
```

I think this is related to the change mentioned in the notes here https://bakery.cakephp.org/2019/12/28/cakephp_388_released.html

> When generating URLs, routes with controller and action placeholders no longer strip these these keys when they are undefined. Instead routes missing the controller and action keys will fail to match.

### What you expected to happen

No BC breaking behavior because of a **patch** release! 😠 Cake is still using https://semver.org/ right?

Is there *any* workaround that can be applied globally to this problem or will I have to manually merge the URL array in **all** places that don't use the controller and action key? This basically means I'll have to do this in all the places that somehow build an URL.

```php
			$url = [
				'plugin' => $request->getParam('plugin'),
				'controller' => $request->getParam('controller'),
				'action' => $request->getParam('action'),
				'?' => array_merge($query, ['char' => $char])
			];
```

**Comment by dereuromark** [comment]
This is only happening inside a test case, correct?
I had to fix up my tests here as well. If you just set some value for the controller the test case will work again fine.

**Comment by burzum** [comment]
So far it seems to happen only in the tests but how can I trust the framework and app if we have failing tests? This is also hard to sell to my boss for obvious reasons.

**Comment by markstory** [comment]
The previous behavior was not correct and was masked by incorrect tests. If you are unable to fix your tests you can pin to 3.8.7.
